src/cycles/simple_cycles.c

```diff
@@ -0,0 +1,506 @@
+    igraph_vector_int_t vertex_stack;
+
+    /* Stack in which the edges of the current cycle are pushed */
+    igraph_vector_int_t edge_stack;
+
+    /* Boolean vector indicating which vertices are blocked */
+    igraph_vector_bool_t blocked;
+
+    /* Whether the graph is directed */
+    igraph_bool_t directed;
+
+    /* Hashes of the vertices of found cycles (for filtering in undirected graphs) */
+    igraph_vector_int_t found_cycles_vertex_hashes;
+
+    /* Hashes of the edges of found cycles (for filtering in undirected graphs) */
+    igraph_vector_int_t found_cycles_edge_hashes;
+} igraph_simple_cycle_search_state_t;
+
+/**
+ * \experimental
+ *
+ * A custom function to prevent double results in the search when using undirected graphs
+ */
+
+static igraph_bool_t igraph_i_cycle_has_been_found_already(
+    igraph_simple_cycle_search_state_t *state,
+    igraph_vector_int_list_t *v_results, igraph_vector_int_list_t *e_results,
+    igraph_vector_int_t *v_res, igraph_vector_int_t *e_res,
+    unsigned long long vertex_hash, unsigned long long edge_hash) {
+    if (igraph_vector_int_size(v_res) < 2) {
+        return false;
+    }
+
+    // TODO: improve performance of this by using some sort of hash table lookup
+    igraph_integer_t pos;
+    if (!igraph_vector_int_search(&state->found_cycles_edge_hashes, 0, edge_hash, &pos)) {
+        return false;
+    }
+    // In principle, it should not be possible to have the same edges but different vertices
+    // Yet this is just another safeguard against a possible (edge) hash collision
+    while (VECTOR(state->found_cycles_vertex_hashes)[pos] != vertex_hash && igraph_vector_int_size(&VECTOR(*v_results)[pos]) != igraph_vector_int_size(v_res)) {
+        if (!igraph_vector_int_search(&state->found_cycles_edge_hashes, pos + 1, edge_hash, &pos)) {
+            return false;
+        }
+    }
+
+    // if all of edge_hash, vertex_hash and cycle size match, we assume we got a duplicate
+    return true;
+}
+
+/**
+ * \experimental
+ *
+ * A hashing algorithm
+ */
+static unsigned long long igraph_i_hash_vector_int(igraph_vector_int_t *vec) {
+
+    unsigned long long hash_product = 1.;
+    unsigned long long hash_sum = 0;
+    unsigned int hash_xor = 0;
+
+    for (igraph_integer_t i = 0; i < igraph_vector_int_size(vec); ++i) {
+        hash_product *= VECTOR(*vec)[i];
+        hash_sum += VECTOR(*vec)[i];
+        hash_xor ^= VECTOR(*vec)[i];
+    }
+
+    return hash_product + hash_sum + ((unsigned long long)hash_xor << 32);
+}
+
+/**
+ * \experimental
+ *
+ * The implementation of procedure UNBLOCK from Johnson's paper
+ *
+ */
+
+static igraph_error_t igraph_i_simple_cycles_unblock(
+    igraph_simple_cycle_search_state_t *state, igraph_integer_t u) {
+    // TODO: introduce stack for w & neis in order to reduce the number of iterations.
+    igraph_vector_int_t *neis;
+    igraph_integer_t w;
+    igraph_stack_int_t u_stack;
+    igraph_stack_int_init(&u_stack, 0);
```

Comment:
Please use `IGRAPH_STACK_INT_INIT_FINALLY` here and add the corresponding `FINALLY_CLEAN` after destroying the stack.

There might be other places where this is missing, I won't leave additional comments now.